karpenter-operator/controllers/karpenter/karpenter_controller_test.go

```diff
@@ -227,3 +231,202 @@ func TestGetUserDataSecret(t *testing.T) {
+			namespace: "test-namespace",
+			hcp: &hyperv1.HostedControlPlane{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-hcp",
+					Namespace: "test-namespace",
+					DeletionTimestamp: &metav1.Time{
+						Time: time.Now(),
+					},
+					Finalizers: []string{
+						karpenterFinalizer,
+						"some-other-finalizer",
+					},
+				},
+			},
+			objects: func() []client.Object {
+				nodepool := &unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"apiVersion": "karpenter.sh/v1",
+						"kind":       "NodePool",
+						"metadata": map[string]interface{}{
+							"name":      "test-nodepool-1",
+							"namespace": "test-namespace",
+						},
+					},
+				}
+				nodepool.SetFinalizers([]string{"some-finalizer"}) // this prevents the nodepool from being deleted
+				return []client.Object{nodepool}
+			}(),
+			expectedNodePools:                   1,
+			eventuallyKarpenterFinalizerRemoved: false,
+		},
+		{
+			name:      "when hcp is deleted, should not remove karpenter finalizer if karpenter NodeClaims still exist",
+			namespace: "test-namespace",
+			hcp: &hyperv1.HostedControlPlane{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-hcp",
+					Namespace: "test-namespace",
+					DeletionTimestamp: &metav1.Time{
+						Time: time.Now(),
+					},
+					Finalizers: []string{
+						karpenterFinalizer,
+						"some-other-finalizer",
+					},
+				},
+			},
+			objects: []client.Object{
+				&unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"apiVersion": "karpenter.sh/v1",
+						"kind":       "NodePool",
+						"metadata": map[string]interface{}{
+							"name":      "test-nodepool-1",
+							"namespace": "test-namespace",
+						},
+					},
+				},
+				&unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"apiVersion": "karpenter.sh/v1",
+						"kind":       "NodeClaim",
+						"metadata": map[string]interface{}{
+							"name":      "test-nodeclaim-1",
+							"namespace": "test-namespace",
+						},
+					},
+				},
+			},
+			expectedNodePools:                   0,
+			eventuallyKarpenterFinalizerRemoved: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			fakeManagementClient := fake.NewClientBuilder().
+				WithScheme(scheme).
+				WithObjects(tc.hcp).
+				Build()
+
+			fakeGuestClient := fake.NewClientBuilder().
+				WithScheme(scheme).
+				WithObjects(tc.objects...).
+				Build()
+
+			r := &Reconciler{
+				ManagementClient: fakeManagementClient,
+				GuestClient:      fakeGuestClient,
+				Namespace:        tc.namespace,
+			}
+
+			ctx := log.IntoContext(context.Background(), testr.New(t))
+
+			_, err := r.Reconcile(ctx, ctrl.Request{NamespacedName: client.ObjectKeyFromObject(tc.hcp)})
+
+			g.Expect(err).NotTo(HaveOccurred())
+
+			// reconcile "enough" times and check if karpenter finalizer on the hcp still exists
+			numReconciles := 3
```

Comment:
Nice test! For the unit test scenario, shouldn't this just need one single reconcile?

Reply:
Yes you are right, fixed, thanks!